### PR TITLE
feat: sequence completion (SIDX/RIDX)

### PR DESCRIPTION
**Pull request by lukaszkolodziejczyk**
Changes:
- the position is encoded with `sidx` and `ridx` columns and they are fixed to the beginning in columns order (both in training and generation)
- each target sequence is padded with additional row to handle 0 sequence lengths
- `ridx` is removed from input to
    - history LSTM
    - column_embedder
- loss changes:
    - `sidx` is excluded from loss as before
    - <s>`padding_mask` is applied to all other columns, in particular `ridx` (on main, `slen` column is masked beyond first step)</s> EDIT: see follow up question 2 for more precise description.
    - normalise batch loss by dividing by total sequence steps in batch
- introduction of seed for sequential tables (`seed_data` argument to `generate`)
    - the seed requires `tgt_context_key` column
    - batches contain partial sequences of the same length (== potentially more batches depending on seed structure)
- automated test for sequential seed was added
- misc changes: `continue_mask` -> `include_mask` name change; batch size is now reduced in generation loop already after single sequence has finished

This aims to be backward compatible during generation, by handling both `sidx/ridx` and `slen/sidx/sdec` cases.

To test this, it's useful to use `simulations.py` script, get it with: `git checkout 8d914950ffde03f0eca0a5a12b469a4a03fe7e3b -- simulations.py`.

Follow up questions and answers:

1. What (if any) is the ridx col input into?

> When ridx is digit-encoded, the sub column `ridx_E{i}` will see the previous sub columns `ridx_E{N}` to `ridx_E{i+1}`

2. How does the loss for ridx work (masking-wise)?

> We pad all the sequences by 1 step, so all of them will have `ridx=0` at the last step.
The mask of `ridx` therefore covers one more step than the mask of other regular columns, so that it can learn to stop as well as the 0-len sequences. `ridx_mask = torch.nn.functional.pad(padding_mask, (1, 0), value=1)[:, :-1]`

### CODE REVIEW

File: mostlyai/engine/_tabular/generation.py [hunk 1]
```diff
@@ -670,26 +683,47 @@ def generate(
...
         _LOG.info(f"{enable_flexible_generation=}")
 
+        # handle different approaches to sequence modeling (backwards compatibility)
+        has_ridx = has_slen = has_sdec = False
+        if is_sequential:
+            if isinstance(model_units, dict):
+                has_ridx = any(RIDX_SUB_COLUMN_PREFIX in k for k in model_units.keys())
+                has_slen = any(SLEN_SUB_COLUMN_PREFIX in k for k in model_units.keys())
+                has_sdec = any(SDEC_SUB_COLUMN_PREFIX in k for k in model_units.keys())
+            else:
```
Comment (shuangwu5) [inline mostlyai/engine/_tabular/generation.py:702]: In what scenario would `model_units` not be a dict? (i.e., when would `model_configs.get("model_units")` be empty?)

File: mostlyai/engine/_tabular/generation.py [hunk 2]
```diff
@@ -618,9 +611,29 @@ def decode_buffered_samples(
...
+            # overwrite first steps of each sequence in synthetic data with values from seed data
+            df_syn["_SEQ_IDX"] = df_syn.groupby(tgt_context_key).cumcount()
+            df_seed["_SEQ_IDX"] = df_seed.groupby(tgt_context_key).cumcount()
+            # df_overwrite is a dataframe with the same shape as df_syn, but with the seed values for the first steps of each sequence
+            df_overwrite = pd.merge(
+                df_syn[[tgt_context_key, "_SEQ_IDX"]].copy(),
+                df_seed.assign(_SEED_IDX=df_seed.index),
+                on=[tgt_context_key, "_SEQ_IDX"],
+                how="left",
+            )
```
Comment (shuangwu5) [inline mostlyai/engine/_tabular/generation.py:628]: Maybe slightly simpler to use `indicator=True` and then check the value of `_merge` column, instead of adding a special `_SEED_IDX` column
(ref: https://pandas.pydata.org/docs/reference/api/pandas.merge.html)

File: mostlyai/engine/_common.py [hunk 3]
```diff
@@ -512,72 +514,56 @@ def skip_if_error_wrapper(*args, **kwargs) -> Any:
...
-        return syn
-
-    # use SIDX and SLEN to determine sequence length
-    syn[SIDX_SUB_COLUMN_PREFIX] = decode_slen_sidx_sdec(syn, seq_len_max, prefix=SIDX_SUB_COLUMN_PREFIX)
-    syn[SLEN_SUB_COLUMN_PREFIX] = decode_slen_sidx_sdec(syn, seq_len_max, prefix=SLEN_SUB_COLUMN_PREFIX)
-    # ensure that seq_len_min is respected
-    syn[SLEN_SUB_COLUMN_PREFIX] = np.maximum(seq_len_min, syn[SLEN_SUB_COLUMN_PREFIX])
-    syn = syn[syn[SIDX_SUB_COLUMN_PREFIX] < syn[SLEN_SUB_COLUMN_PREFIX]].reset_index(drop=True)
-    # discarded padded context rows, ie where context key has been set to None
-    syn = syn.dropna(subset=[tgt_context_key])
```
[the remark this answers is not in the document]
Comment (lukaszkolodziejczyk) [reply mostlyai/engine/_common.py:572]: No, in my opinion, this one is not relevant anymore. I think we dragged this from the times when the engine was implemented with TF and each batch had to strictly be of the same size.

File: mostlyai/engine/_tabular/generation.py (hunk 1 above)
Comment (shuangwu5) [reply mostlyai/engine/_tabular/generation.py:702]: Ok, maybe having a comment would help 🙂

File: mostlyai/engine/_tabular/training.py [hunk 4]
```diff
@@ -264,34 +266,30 @@ def _calculate_sample_losses(
...
+            elif col in ridx_cols:
+                mask = ridx_mask
             else:
-                # mask out paddings
-                mask = slen_mask
-
+                mask = padding_mask
             column_loss = criterion(output[col].transpose(1, 2), data[col].squeeze(2))
-            masked_loss = torch.sum(column_loss * mask, dim=1) / torch.clamp(torch.sum(mask, dim=1), min=1)
+            masked_loss = torch.sum(column_loss * mask, dim=1) / torch.clamp(torch.sum(mask >= 1), min=1)
```
Comment (andre-mostly) [inline mostlyai/engine/_tabular/training.py:292]: Any reason not just to sum the mask? Do we have any other value than 0 and 1?

File: mostlyai/engine/_tabular/argn.py [hunk 5]
```diff
@@ -214,9 +215,16 @@ def __init__(
         self.embedders = nn.ModuleDict()
 
         # embedding layers for each sub column defined in cardinalities
+        last_ridx_sub_col = next(
+            (sub_col for sub_col in reversed(self.cardinalities) if sub_col.startswith(RIDX_SUB_COLUMN_PREFIX)), None
+        )
         for sub_col, dim_input in self.cardinalities.items():
             dim_output = _embedding_heuristic(id=self.id(sub_col), model_size=model_size, dim_input=dim_input)
             embedder = nn.Embedding(num_embeddings=dim_input, embedding_dim=dim_output, device=device)
+            # the embeddings of the last ridx sub column are never used
```
[the remark this answers is not in the document]
Comment (shuangwu5) [reply mostlyai/engine/_tabular/argn.py:224]: Yes this holds for both cases.
The embedding of last ridx sub column (those with suffix `_cat` for categorical encoding and `_E0` for digit-encoding) are never fed to any layer -> no gradients -> opacus complaints about it if not with `requires_grad=False`

File: mostlyai/engine/_tabular/training.py [hunk 6]
```diff
@@ -264,34 +266,30 @@ def _calculate_sample_losses(
...
+
+        # mask for all the columns other than sidx and ridx
+        padding_mask = torch.zeros_like(data[ridx_cols[0]], dtype=torch.int64)
+        for ridx_col in ridx_cols:
+            padding_mask |= data[ridx_col] != 0  # mask loss for padded rows, which have RIDX=0
+        padding_mask = padding_mask.squeeze(-1)
+        # mask for ridx columns: this takes the sequence padding into account to learn the stopping with ridx=0
+        ridx_mask = torch.nn.functional.pad(padding_mask, (1, 0), value=1)[:, :-1]
+        # mask for sidx columns
+        sidx_mask = torch.zeros_like(padding_mask, dtype=torch.int64)
```
Comment (andre-mostly) [inline mostlyai/engine/_tabular/training.py:280]: torch.int64 should be inherited from padding_mask

File: mostlyai/engine/_tabular/argn.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (andre-mostly) [reply mostlyai/engine/_tabular/argn.py:224]: I see that after our slack chat, thank you.

File: mostlyai/engine/_tabular/training.py (hunk 4 above)
Comment (lukaszkolodziejczyk) [reply mostlyai/engine/_tabular/training.py:292]: Yeah, I think it's worth keeping it as is, it's a bit more future proof I feel.

**Review by shuangwu5 — APPROVED** [review]
